### PR TITLE
Add administrative webhooks

### PR DESCRIPTION
**Pull request by Gargron**
Webhooks would benefit automation through the moderation API by notifying applications about system events in real-time. They would also enable integrations with chat apps like Discord, IRC and Slack, helping moderator coordination.

Events currently supported:

- `report.created`
- `account.created`

The `X-Hub-Signature` header adopted from the WebSub spec can be optionally used to verify that the payloads are authentic.

The payload delivered to the webhook has the form of:

```js
{
  "event": "report.created",
  "created_at": "2022-06-08T12:28:31.000Z",
  "object": {
    // The report entity would be here
  }
}
```

Mind that `created_at` at the top-level refers to the event itself, and not the `object`.

___

Screenshot 2022-06-08 at 00-38-37 New webhook - Mastodon (Dev)
Screenshot 2022-06-08 at 00-38-51 Webhooks - Mastodon (Dev)
Screenshot 2022-06-08 at 00-39-05 Webhooks - Mastodon (Dev)

___

This project was funded through the [NGI0 Discovery](https://nlnet.nl/discovery) Fund, a fund established by [NLnet](https://nlnet.nl/) with financial support from the European Commission's [Next Generation Internet](https://ngi.eu/) programme, under the aegis of DG Communications Networks, Content and Technology under grant agreement No 825322.

### CODE REVIEW

**Review by ClearlyClaire — APPROVED** [review]
Looks good to me overall, though I made some minor inline comments.

The list of implemented events is pretty limited, but there is no issue with that being extended in subsequent PRs.

**Comment by ashleyhull-versent** [comment]
Brainstormed additional events:
* report.resolved
* report.markedassensitive
* report.deleted
* account.warn
* account.freeze
* account.forcesensitive
* account.limit
* account.suspend
* account.memoriam
* instance.silenced
* instance.suspended
* announcements.enabled

Other points of interest
* sidekiq.queue.latency (over X value)

File: app/services/base_service.rb [hunk 1]
```diff
@@ -5,4 +5,8 @@ class BaseService
   include ActionView::Helpers::SanitizeHelper
 
   include RoutingHelper
+
+  def call(*)
+    raise NotImplementedError
+  end
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/services/base_service.rb:11]: Just code quality, as I see it, defining the role of `BaseService` more clearly...

**Review by ClearlyClaire — APPROVED** [review]

**Comment by wioxjk** [comment]
I tried to get it to work with the Webhook-functionality provided by t2boit.io - in order to get notifications into a matrix-room.
However, it did not work. The suspicion is that this webhook is not "slack compatible" - is that right?

**Comment by Saiv46** [comment]
If you wanted to make webhooks and stumbled upon this PR, since this is undocumented:

> Webhooks are HTTP URLs for sending data of selected events, such as user registration, so it can be processed by other service.
> 
> When triggered by some event, instance will make HTTP POST request to configured URL with body with `application/json` encoding.
> 
> ```json
> {
>   "event": "account.created",
>   "created_at": "2023-01-04T17:37:53.374Z",
>   "object": // Admin::Account
> }
> ```
> 
> JSON body includes event name (`event`), event creation time (`created_at`) and object that is related to event (in case of `account.created` - an instance of [Admin::Account](https://docs.joinmastodon.org/entities/Admin_Account/)).
> 
> Recieved response body will be ignored and only status code used:
> - 2XX-3XX - Request successful.
> - 401, 408, 429 - Request will be retried (max 16).
> - 4XX-5XX - Request failed, no further attempts.
> 
> To be able to [validate request payloads](https://www.w3.org/TR/websub/#authenticated-content-distribution), `X-Hub-Signature` header can be used, which will be in a form of `method=signature` where `method` is one of the recognized algorithm names (SHA1, SHA256, SHA384, SHA512) and `signature` is the hexadecimal representation of the signature. Signature is computed using request payload as data and *signing secret* as key.
>

**Comment by kevquirk** [comment]
Is there a facility to reference objects like report number from a webhook?

**Comment by carrotcypher** [comment]
This lacking documentation was an expected head scratcher for a moment. For the benefit of anyone else coming to this page and scratching their heads, here's how we go it working for `Mastodon reports > Mattermost mod channel notifications` for us at Fosstodon.org.


### 1) Create the new webhook endpoint on Mattermost

image


### 2) Create the new webhook event on Mastodon

image

_For our needs, we only wanted it triggered for the `report.created` event._

### 3) Change the "Payload Template" of the Mastodon webhook event to:

`{ "text": "Here is the text to send, and you can reference values with {{object.value}} depending on its location in the JSON tree (see below)" }`

_Refer to the Reports API JSON schema for individual variables you are looking for: https://docs.joinmastodon.org/methods/reports/_

Here is the example schema reference there:

```{
  "id": "48914",
  "action_taken": false,
  "action_taken_at": null,
  "category": "spam",
  "comment": "Spam account",
  "forwarded": false,
  "created_at": "2022-08-25T09:56:16.763Z",
  "status_ids": [
    "108882889550545820"
  ],
  "rule_ids": null,
  "target_account": {
    "id": "108366849347798387",
    "username": "Baluke",
    "acct": "Baluke",
    "display_name": "Baluke Dental Studios",
    "locked": false,
    "bot": false,
    "discoverable": false,
    "group": false,
    "created_at": "2022-05-26T00:00:00.000Z",
    "note": "<p>Baluke Dental Studios is a full service dental lab offering fabrication, staining, and digital services. Advanced technologies and a meticulous process ensure reduced chair time, lower costs, and better patient outcomes with beautiful smiles. Talk to a representative today.</p><p><a href=\"https://baluke.com/\" target=\"_blank\" rel=\"nofollow noopener noreferrer\"><span class=\"invisible\">https://</span><span class=\"\">baluke.com/</span><span class=\"invisible\"></span></a></p>",
    "url": "https://mastodon.social/@Baluke",
    "avatar": "https://files.mastodon.social/accounts/avatars/108/366/849/347/798/387/original/dbcfe99ed5def0f4.png",
    "avatar_static": "https://files.mastodon.social/accounts/avatars/108/366/849/347/798/387/original/dbcfe99ed5def0f4.png",
    "header": "https://static-cdn.mastodon.social/headers/original/missing.png",
    "header_static": "https://static-cdn.mastodon.social/headers/original/missing.png",
    "followers_count": 0,
    "following_count": 0,
    "statuses_count": 38,
    "last_status_at": "2022-08-25",
    "emojis": [],
    "fields": []
  }
}
```

All of the above JSON is contained in the `object` array, so if we want to refer to any of it we need to use `{{object.value}}`. 

For our case, we wanted the following message in the Mattermost channel on each new report:

> @channel report [10415](https://linktoreporthere) was just filed for the user [BobIsACat](https://linktouserprofile)

To achieve that, we use the fields `object.id` and `object.target_account.url` as follows:

```
{ "text": "@channel report [{{object.id}}](https://ourmastadoninstance.com/reports/{{object.id}}) was just filed for the user [{{object.target_account.url}}]({{object.target_account.url}})" }
```

Hope this was helpful!

**Comment by carrotcypher** [comment]
@Gargron is there also an event for when a mod leaves a note on the report? 


Alternatively, are there plans for any incoming webhooks so that moderation could be done from a third party site/service (for example, in our case we would want to add a moderator note to the report from inside a Mattermost channel)?

**Comment by ClearlyClaire** [comment]
`report.updated` is triggered when the report itself is modified, but not when moderation notes are left on it. There is currently no webhook event for that.

**Pull request merged**